python/cuml/decomposition/incremental_pca.py

```diff
@@ -0,0 +1,536 @@
+
+    Parameters
+    ----------
+    X : scipy.sparse or cupy.sparse object
+        A sparse input
+    Returns
+    -------
+    X : The input converted to a cupy.sparse.csr_matrix object
+    """
+
+    acceptable_dtypes = ('float32', 'float64')
+    acceptable_cupy_sparse_formats = \
+        (cp.sparse.csr_matrix, cp.sparse.coo_matrix)
+
+    if X.dtype not in acceptable_dtypes:
+        raise TypeError("Expected input to be of type float32 or float64."
+                        " Received %s" % X.dtype)
+    if scipy.sparse.issparse(X):
+        return cp.sparse.csr_matrix(X)
+    elif cp.sparse.issparse(X):
+        if not isinstance(X, acceptable_cupy_sparse_formats):
+            raise TypeError("Expected input to be of type"
+                            " cupy.sparse.csr_matrix or"
+                            " cupy.sparse.csc_matrix. Received %s"
+                            % type(X))
+        else:
+            return X
+
+
+def _gen_batches(n, batch_size, min_batch_size=0):
+    """
+    Generator to create slices containing batch_size elements, from 0 to n.
+    The last slice may contain less than batch_size elements, when batch_size
+    does not divide n.
+
+    Parameters
+    ----------
+    n : int
+    batch_size : int
+        Number of element in each batch
+    min_batch_size : int, default=0
+        Minimum batch size to produce.
+    Yields
+    ------
+    slice of batch_size elements
+    """
+
+    if not isinstance(batch_size, numbers.Integral):
+        raise TypeError("gen_batches got batch_size=%s, must be an"
+                        " integer" % batch_size)
+    if batch_size <= 0:
+        raise ValueError("gen_batches got batch_size=%s, must be"
+                         " positive" % batch_size)
+    start = 0
+    for _ in range(int(n // batch_size)):
+        end = start + batch_size
+        if end + min_batch_size > n:
+            continue
+        yield slice(start, end)
+        start = end
+    if start < n:
+        yield slice(start, n)
+
+
+def _safe_accumulator_op(op, x, *args, **kwargs):
+    """
+    This function provides numpy accumulator functions with a float64 dtype
+    when used on a floating point input. This prevents accumulator overflow on
+    smaller floating point dtypes.
+    Parameters
+    ----------
+    op : function
+        A cupy accumulator function such as cp.mean or cp.sum
+    x : cupy array
+        A numpy array to apply the accumulator function
+    *args : positional arguments
+        Positional arguments passed to the accumulator function after the
+        input x
+    **kwargs : keyword arguments
+        Keyword arguments passed to the accumulator function
+    Returns
+    -------
+    result : The output of the accumulator function passed to this function
+    """
+
+    if cp.issubdtype(x.dtype, cp.floating) and x.dtype.itemsize < 8:
+        result = op(x, *args, **kwargs, dtype=cp.float64).astype(cp.float32)
```

Comment:
The pytests don't seem to be affected, but I wonder if they are enough to create a test that will test whether we need to do those operations in double precision. What do you think?

Reply:
It's ok to change this in a follow-on PR, but I think it would be ideal to have an option to use safe accumulator or 32-bit accumlator for testing. Unless it makes very little perf difference.

Reply:
Perf difference, in terms of speed, have not been significant in the testing that Corey and I have done on V100 and Quadro RTX 8000 respectively. The critical section of the code is in the SVD and I would wager over 95% of the time is spent there on an average. Perf difference, however, in terms of memory is probably significant. Haven't been able to create a test dataset though which shows why that `float64` cast is necessary